File: Modules/DeckFeature/Sources/TurntablePhysics.swift
import Foundation

struct TurntablePhysics {
    var platterPosition: Double
    var angularVelocity: Double
    let inertia: Double
    let damping: Double

    init(
        platterPosition: Double = 0,
        angularVelocity: Double = 0,
        inertia: Double = 0.12,
        damping: Double = 3.5
    ) {
        self.platterPosition = platterPosition
        self.angularVelocity = angularVelocity
        self.inertia = max(inertia, 0.0001)
        self.damping = max(damping, 0)
    }

    mutating func reset(position: Double = 0) {
        platterPosition = normalizedAngle(position)
        angularVelocity = 0
    }

    mutating func applyAngularDrag(deltaAngle: Double, deltaTime: TimeInterval) {
        let safeDeltaTime = max(deltaTime, 0.001)
        platterPosition = normalizedAngle(platterPosition + deltaAngle)
        angularVelocity = deltaAngle / safeDeltaTime
    }

    mutating func step(deltaTime: TimeInterval, driveAngularVelocity: Double?) {
        let safeDeltaTime = min(max(deltaTime, 0), 0.1)
        guard safeDeltaTime > 0 else {
            return
        }

        if let targetVelocity = driveAngularVelocity {
            let blend = min(1, safeDeltaTime / inertia)
            angularVelocity += (targetVelocity - angularVelocity) * blend
        } else {
            angularVelocity *= exp(-damping * safeDeltaTime)
        }

        platterPosition = normalizedAngle(platterPosition + (angularVelocity * safeDeltaTime))
    }

    private func normalizedAngle(_ angle: Double) -> Double {
        let twoPi = 2.0 * Double.pi
        var normalized = angle.truncatingRemainder(dividingBy: twoPi)
        if normalized < 0 {
            normalized += twoPi
        }
        return normalized
    }
}

File: Projects/MixerWeb/packages/mixer-domain/src/turntablePhysics.ts
export type TurntablePhysicsState = {
  platterPositionRadians: number;
  angularVelocity: number;
  inertia: number;
  damping: number;
};

export function createTurntablePhysicsState(
  overrides: Partial<TurntablePhysicsState> = {}
): TurntablePhysicsState {
  return {
    platterPositionRadians: 0,
    angularVelocity: 0,
    inertia: 0.12,
    damping: 3.5,
    ...overrides
  };
}

export function stepTurntablePhysics(
  state: TurntablePhysicsState,
  deltaTime: number,
  driveAngularVelocity?: number
): TurntablePhysicsState {
  const safeDeltaTime = Math.min(Math.max(deltaTime, 0), 0.1);
  if (safeDeltaTime <= 0) {
    return state;
  }

  let angularVelocity = state.angularVelocity;
  if (typeof driveAngularVelocity === "number") {
    const blend = Math.min(1, safeDeltaTime / Math.max(state.inertia, 0.0001));
    angularVelocity += (driveAngularVelocity - angularVelocity) * blend;
  } else {
    angularVelocity *= Math.exp(-Math.max(state.damping, 0) * safeDeltaTime);
  }

  return {
    ...state,
    angularVelocity,
    platterPositionRadians: normalizeAngle(
      state.platterPositionRadians + (angularVelocity * safeDeltaTime)
    )
  };
}

export function applyTurntableDrag(
  state: TurntablePhysicsState,
  deltaAngle: number,
  deltaTime: number
): TurntablePhysicsState {
  const safeDeltaTime = Math.max(deltaTime, 0.001);
  return {
    ...state,
    platterPositionRadians: normalizeAngle(state.platterPositionRadians + deltaAngle),
    angularVelocity: deltaAngle / safeDeltaTime
  };
}

function normalizeAngle(angle: number): number {
  const fullRotation = Math.PI * 2;
  let normalized = angle % fullRotation;
  if (normalized < 0) {
    normalized += fullRotation;
  }
  return normalized;
}
